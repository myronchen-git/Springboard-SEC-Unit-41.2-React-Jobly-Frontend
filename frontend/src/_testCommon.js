const companies = Object.freeze([
  Object.freeze({
    handle: 'comp1',
    name: 'Company 1',
    description: 'Company 1 description.',
    numEmployees: 12,
    logoUrl: 'logo1.png',
  }),
  Object.freeze({
    handle: 'comp2',
    name: 'Company 2',
    description: 'Company 2 description.',
    numEmployees: 123,
    logoUrl: 'logo2.png',
  }),
  Object.freeze({
    handle: 'comp3',
    name: 'Company 3',
    description: 'Company 3 description.',
    numEmployees: 3,
  }),
  Object.freeze({
    handle: 'comp4',
    name: 'Company 4',
    description: 'Company 4 description.',
    logoUrl: 'logo4.png',
  }),
]);

const jobs = Object.freeze([
  Object.freeze({
    id: '1',
    title: 'Job 1',
    salary: 1,
    equity: 0.1,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
  Object.freeze({
    id: '2',
    title: 'Job 2',
    salary: 2,
    equity: 0.2,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
]);

const comp1Jobs = [];
[jobs[0], jobs[1]].forEach((job) => {
  const condensedJob = { ...job };
  delete condensedJob.companyHandle;
  delete condensedJob.companyName;
  comp1Jobs.push(Object.freeze(condensedJob));
});

const companyDetails = Object.freeze({
  handle: 'comp1',
  name: 'Company 1',
  description: 'Company 1 description.',
  numEmployees: 12,
  logoUrl: 'logo1.png',
  jobs: Object.freeze(comp1Jobs),
});

// ==================================================

export { companies, companyDetails, jobs };
