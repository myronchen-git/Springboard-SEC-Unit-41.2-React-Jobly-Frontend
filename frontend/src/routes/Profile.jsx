import PropTypes from 'prop-types';
import { useState } from 'react';

import JoblyApi from '../../../api.js';

// ==================================================

/**
 * Displays a form to update and updates a user's profile.  Currently only
 * allows updating first name, last name, and email.
 *
 * @param {Object} props - React component properties.
 * @param {Object} props.user - The user Object from App.
 * @param {Function} props.setUser - Sets the new user state value in App.
 */
function Profile({ user, setUser }) {
  const initialFormData = {
    firstName: user.firstName,
    lastName: user.lastName,
    email: user.email,
  };
  const [formData, setFormData] = useState(initialFormData);
  const [feedbackMessage, setFeedbackMessage] = useState(null);

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();

    let updatedUser;
    try {
      updatedUser = await JoblyApi.patchUser(user.username, formData);
    } catch (err) {
      setFeedbackMessage(err[0]);
      return;
    }

    setFeedbackMessage('Profile Updated');
    setUser(updatedUser);
  }

  return (
    <main className="Profile">
      <header>
        <h1>Profile</h1>
      </header>
      <form className="Profile__form" onSubmit={handleSubmit}>
        <div>
          <label htmlFor="Profile__input-firstName">First Name</label>
          <input
            id="Profile__input-firstName"
            type="text"
            name="firstName"
            value={formData.firstName}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="Profile__input-lastName">Last Name</label>
          <input
            id="Profile__input-lastName"
            type="text"
            name="lastName"
            value={formData.lastName}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="Profile__input-email">Email</label>
          <input
            id="Profile__input-email"
            type="email"
            name="email"
            value={formData.email}
            onChange={handleChange}
          />
        </div>
        {feedbackMessage && (
          <div className="Profile__feedback">
            <p>{feedbackMessage}</p>
          </div>
        )}
        <button type="submit">Submit</button>
      </form>
    </main>
  );
}

// ==================================================

Profile.propTypes = {
  user: PropTypes.object.isRequired,
  setUser: PropTypes.func.isRequired,
};

export default Profile;
