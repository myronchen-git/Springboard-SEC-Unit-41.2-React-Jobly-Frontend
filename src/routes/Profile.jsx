import PropTypes from 'prop-types';
import { useState } from 'react';
import {
  Alert,
  Button,
  Card,
  CardBody,
  CardTitle,
  Form,
  FormGroup,
  Input,
  Label,
} from 'reactstrap';

import JoblyApi from '../api.js';

// ==================================================

/**
 * Displays a form to update and updates a user's profile.  Currently only
 * allows updating first name, last name, and email.
 *
 * @param {Object} props - React component properties.
 * @param {Object} props.user - The user Object from App.
 * @param {Function} props.setUser - Sets the new user state value in App.
 */
function Profile({ user, setUser }) {
  const initialFormData = {
    firstName: user.firstName,
    lastName: user.lastName,
    email: user.email,
  };
  const [formData, setFormData] = useState(initialFormData);
  const [feedbackMessage, setFeedbackMessage] = useState({});

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();

    let updatedUser;
    try {
      updatedUser = await JoblyApi.patchUser(user.username, formData);
    } catch (err) {
      setFeedbackMessage({ type: 'danger', message: err[0] });
      return;
    }

    setFeedbackMessage({ type: 'success', message: 'Profile Updated' });
    setUser(updatedUser);
  }

  return (
    <main className="Profile">
      <Card>
        <CardBody>
          <CardTitle tag={'h2'}>Profile</CardTitle>
          <Form className="Profile__form" onSubmit={handleSubmit}>
            <FormGroup className="text-start">
              <Label htmlFor="Profile__input-firstName">
                <b>First Name</b>
              </Label>
              <Input
                id="Profile__input-firstName"
                type="text"
                name="firstName"
                value={formData.firstName}
                onChange={handleChange}
              />
            </FormGroup>
            <FormGroup className="text-start">
              <Label htmlFor="Profile__input-lastName">
                <b>Last Name</b>
              </Label>
              <Input
                id="Profile__input-lastName"
                type="text"
                name="lastName"
                value={formData.lastName}
                onChange={handleChange}
              />
            </FormGroup>
            <FormGroup className="text-start">
              <Label htmlFor="Profile__input-email">
                <b>Email</b>
              </Label>
              <Input
                id="Profile__input-email"
                type="email"
                name="email"
                value={formData.email}
                onChange={handleChange}
              />
            </FormGroup>
            {feedbackMessage.message && (
              <Alert color={feedbackMessage.type}>
                {feedbackMessage.message}
              </Alert>
            )}
            <Button color="light" type="submit">
              Submit
            </Button>
          </Form>
        </CardBody>
      </Card>
    </main>
  );
}

// ==================================================

Profile.propTypes = {
  user: PropTypes.object.isRequired,
  setUser: PropTypes.func.isRequired,
};

export default Profile;
