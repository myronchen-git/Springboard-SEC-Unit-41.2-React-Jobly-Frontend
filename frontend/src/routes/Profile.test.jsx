import { act, fireEvent, render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import JoblyApi from '../../../api';
import { userData } from '../_testCommon.js';
import Profile from './Profile.jsx';

// ==================================================

vi.mock(import('../../../api'), () => {
  const MockJoblyApi = vi.fn();
  MockJoblyApi.patchUser = vi.fn();

  return {
    default: MockJoblyApi,
  };
});

// ==================================================

describe('Profile', () => {
  const mockSetUser = vi.fn();

  const user = { ...userData };
  delete user.applications;
  Object.freeze(user);

  const updatedData = Object.freeze({
    firstName: 'Updated First',
    lastName: 'Updated Last',
    email: 'updated.email@email.com',
  });

  const updatedUser = Object.assign({}, user, updatedData);

  beforeEach(() => {
    JoblyApi.patchUser.mockReset();
    mockSetUser.mockReset();

    JoblyApi.patchUser.mockResolvedValue(updatedUser);
  });

  it('renders.', async () => {
    render(<Profile user={user} setUser={mockSetUser} />);
  });

  it('matches snapshot.', async () => {
    const { asFragment } = render(
      <Profile user={user} setUser={mockSetUser} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('can update form data.', () => {
    // Arrange
    const { container, getByLabelText } = render(
      <Profile user={user} setUser={mockSetUser} />
    );

    const firstNameInput = getByLabelText('First Name');
    const lastNameInput = getByLabelText('Last Name');
    const emailInput = getByLabelText('Email');

    // Act
    fireEvent.change(firstNameInput, {
      target: { value: updatedData.firstName },
    });
    fireEvent.change(lastNameInput, {
      target: { value: updatedData.lastName },
    });
    fireEvent.change(emailInput, { target: { value: updatedData.email } });

    // Assert
    expect(firstNameInput).toHaveValue(updatedData.firstName);
    expect(lastNameInput).toHaveValue(updatedData.lastName);
    expect(emailInput).toHaveValue(updatedData.email);

    const feedbackElement = container.querySelector('.Profile__feedback');
    expect(feedbackElement).not.toBeInTheDocument();

    expect(mockSetUser).not.toHaveBeenCalled();
  });

  it('can submit form data.', async () => {
    // Arrange
    const { getByLabelText, getByText } = render(
      <Profile user={user} setUser={mockSetUser} />
    );

    fireEvent.change(getByLabelText('First Name'), {
      target: { value: updatedData.firstName },
    });
    fireEvent.change(getByLabelText('Last Name'), {
      target: { value: updatedData.lastName },
    });
    fireEvent.change(getByLabelText('Email'), {
      target: { value: updatedData.email },
    });

    // Act
    await act(async () => fireEvent.click(getByText('Submit')));

    // Assert
    expect(getByText('Profile Updated')).toBeVisible();

    expect(JoblyApi.patchUser).toHaveBeenCalledOnce();
    expect(JoblyApi.patchUser).toHaveBeenCalledWith(user.username, updatedData);
    expect(mockSetUser).toHaveBeenCalledOnce();
    expect(mockSetUser).toHaveBeenCalledWith(updatedUser);
  });

  it(
    'displays an error message if there is an issue ' +
      'when updating profile.',
    async () => {
      // Arrange
      const errorMessage = 'error';
      JoblyApi.patchUser.mockRejectedValue([errorMessage]);

      const { getByLabelText, getByText } = render(
        <Profile user={user} setUser={mockSetUser} />
      );

      fireEvent.change(getByLabelText('First Name'), {
        target: { value: updatedData.firstName },
      });
      fireEvent.change(getByLabelText('Last Name'), {
        target: { value: updatedData.lastName },
      });
      fireEvent.change(getByLabelText('Email'), {
        target: { value: updatedData.email },
      });

      // Act
      await act(async () => fireEvent.click(getByText('Submit')));

      // Assert
      expect(getByText(errorMessage)).toBeVisible();

      expect(JoblyApi.patchUser).toHaveBeenCalledOnce();
      expect(JoblyApi.patchUser).toHaveBeenCalledWith(
        user.username,
        updatedData
      );
      expect(mockSetUser).not.toHaveBeenCalled();
    }
  );
});
