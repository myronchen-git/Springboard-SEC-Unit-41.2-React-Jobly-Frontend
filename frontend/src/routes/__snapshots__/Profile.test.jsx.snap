// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="Profile"
  >
    <header>
      <h1>
        Profile
      </h1>
    </header>
    bound HTMLFormElement {
      "0": <input
        id="Profile__input-firstName"
        name="firstName"
        type="text"
        value="First"
      />,
      "1": <input
        id="Profile__input-lastName"
        name="lastName"
        type="text"
        value="Last"
      />,
      "2": <input
        id="Profile__input-email"
        name="email"
        type="email"
        value="email@email.com"
      />,
      "3": <button
        type="submit"
      >
        Submit
      </button>,
    }
  </main>
</DocumentFragment>
`;
